File: src/test/java/de/spries/fleetcommander/model/ai/ComputerPlayerTest.java
package de.spries.fleetcommander.model.ai;

import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

import de.spries.fleetcommander.model.ai.behavior.ProductionStrategy;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mockito;

import de.spries.fleetcommander.model.ai.behavior.BuildingStrategy;
import de.spries.fleetcommander.model.ai.behavior.FleetStrategy;
import de.spries.fleetcommander.model.core.Game;
import de.spries.fleetcommander.model.core.Player.Status;
import de.spries.fleetcommander.model.core.universe.Universe;
import de.spries.fleetcommander.model.facade.PlayerSpecificUniverse;

public class ComputerPlayerTest {

	private ComputerPlayer player;
	private Game game;
	private Universe universe;
	private BuildingStrategy buildingStrategy;
	private FleetStrategy fleetStrategy;
	private ProductionStrategy prodStrategy;

	@Before
	public void setUp() {
		buildingStrategy = mock(BuildingStrategy.class);
		fleetStrategy = mock(FleetStrategy.class);
		prodStrategy = mock(ProductionStrategy.class);
		player = new ComputerPlayer("Computer", buildingStrategy, fleetStrategy, prodStrategy);

		game = mock(Game.class);
		universe = mock(Universe.class);

		doReturn(universe).when(game).getUniverse();
	}

	@Test
	public void isntAHumanPlayer() throws Exception {
		assertThat(player.isHumanPlayer(), is(false));
	}

	@Test
	public void isReadyInitially() throws Exception {
		assertThat(player.getStatus(), is(Status.READY));
	}

	@Test
	public void callsBuildingStrategy() throws Exception {
		player.notifyNewTurn(game);
		verify(buildingStrategy).buildFactories(Mockito.any(PlayerSpecificUniverse.class));
	}

	@Test
	public void callsFleetStrategy() throws Exception {
		player.notifyNewTurn(game);
		verify(fleetStrategy).sendShips(Mockito.any(PlayerSpecificUniverse.class));
	}

	@Test
	public void callsProductionStrategy() throws Exception {
		player.notifyNewTurn(game);
		verify(prodStrategy).updateProductionFocus(Mockito.any(PlayerSpecificUniverse.class), anyInt());
	}

	@Test
	public void endsTurn() {
		player.notifyNewTurn(game);
		verify(game).endTurn(player);
	}

	@Test
	public void stillEndsTurnWhenPlayerHasNoMoreHomePlanet() throws Exception {
		doThrow(RuntimeException.class).when(universe).getHomePlanetOf(player);
		player.notifyNewTurn(game);
		verify(game).endTurn(player);
	}

	@Test
	public void endsTurnIfFleetStrategyThrowsException() throws Exception {
		doThrow(RuntimeException.class).when(fleetStrategy).sendShips(Mockito.any(PlayerSpecificUniverse.class));
		player.notifyNewTurn(game);
		verify(game).endTurn(player);
	}

	@Test
	public void endsTurnIfBuildingStrategyThrowsException() throws Exception {
		doThrow(RuntimeException.class).when(buildingStrategy)
				.buildFactories(Mockito.any(PlayerSpecificUniverse.class));
		player.notifyNewTurn(game);
		verify(game).endTurn(player);
	}

}

File: src/test/kotlin/de/spries/fleetcommander/model/ai/ComputerPlayerTest.kt
package de.spries.fleetcommander.model.ai

import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import de.spries.fleetcommander.model.ai.behavior.BuildingStrategy
import de.spries.fleetcommander.model.ai.behavior.FleetStrategy
import de.spries.fleetcommander.model.ai.behavior.ProductionStrategy
import de.spries.fleetcommander.model.core.Game
import de.spries.fleetcommander.model.core.Player.Status
import de.spries.fleetcommander.model.core.universe.Universe
import org.hamcrest.Matchers.`is`
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito.doThrow
import org.mockito.Mockito.verify

class ComputerPlayerTest {

    private lateinit var player: ComputerPlayer
    private lateinit var game: Game
    private lateinit var universe: Universe
    private lateinit var buildingStrategy: BuildingStrategy
    private lateinit var fleetStrategy: FleetStrategy
    private lateinit var prodStrategy: ProductionStrategy

    @Before
    fun setUp() {
        buildingStrategy = mock()
        fleetStrategy = mock()
        prodStrategy = mock()
        player = ComputerPlayer("Computer", buildingStrategy, fleetStrategy, prodStrategy)

        game = mock()
        universe = mock()
        whenever(game.getUniverse()).thenReturn(universe)
    }

    @Test
    fun isntAHumanPlayer() {
        assertThat(player.isHumanPlayer(), `is`(false))
    }

    @Test
    fun isReadyInitially() {
        assertThat(player.getStatus(), `is`(Status.READY))
    }

    @Test
    fun callsBuildingStrategy() {
        player.notifyNewTurn(game)
        verify(buildingStrategy).buildFactories(any())
    }

    @Test
    fun callsFleetStrategy() {
        player.notifyNewTurn(game)
        verify(fleetStrategy).sendShips(any())
    }

    @Test
    fun callsProductionStrategy() {
        player.notifyNewTurn(game)
        verify(prodStrategy).updateProductionFocus(any(), any())
    }

    @Test
    fun endsTurn() {
        player.notifyNewTurn(game)
        verify(game).endTurn(player)
    }

    @Test

    fun stillEndsTurnWhenPlayerHasNoMoreHomePlanet() {
        whenever(universe.getHomePlanetOf(player)).thenThrow(RuntimeException())
        player.notifyNewTurn(game)
        verify(game).endTurn(player)
    }

    @Test
    fun endsTurnIfFleetStrategyThrowsException() {
        whenever(fleetStrategy.sendShips(any())).thenThrow(RuntimeException())
        player.notifyNewTurn(game)
        verify(game).endTurn(player)
    }

    @Test
    fun endsTurnIfBuildingStrategyThrowsException() {
        doThrow(RuntimeException::class.java).`when`(buildingStrategy)
                .buildFactories(any())
        player.notifyNewTurn(game)
        verify(game).endTurn(player)
    }

}
